### PR TITLE
feat: add support for Yarn

### PR DESCRIPTION
**Pull request by iansu**

### CODE REVIEW

**Comment by iansu** [comment]
Looks like the code has changed a fair bit since I made this change on Saturday. I'll rebase and update.

**Comment by smeijer** [comment]
Sorry about that. :innocent:

File: src/index.js [hunk 1]
```diff
@@ -10,15 +10,23 @@ const write = util.promisify(fs.writeFile);
...
-async function test(pkg) {
+async function test(pkg, isYarn) {
   const spinner = ora(`${pkg} installing`).start();
 
-  await exec(`npm install ${pkg} --no-save --no-audit`);
+  if (!isYarn) {
+    await exec(`npm install ${pkg} --no-save --no-audit`);
+  } else {
+    await exec(`yarn add ${pkg} --no-lockfile`);
+  }
```
Comment (MichaelDeBoey) [inline src/index.js]: Make this positive?

```suggestion
  if (isYarn) {
    await exec(`yarn add ${pkg} --no-lockfile`);
  } else {
    await exec(`npm install ${pkg} --no-save --no-audit`);
  }
```

File: src/index.js [hunk 2]
```diff
@@ -10,15 +10,23 @@ const write = util.promisify(fs.writeFile);
...
 
   spinner.text = `${pkg} testing`;
 
   try {
-    await exec("npm run test");
+    if (!isYarn) {
+      await exec('npm run test');
+    } else {
+      await exec('yarn test');
+    }
```
Comment (MichaelDeBoey) [inline src/index.js]: Same

```suggestion
    if (isYarn) {
      await exec('yarn test');
    } else {
      await exec('npm run test');
    }
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

**Comment by iansu** [comment]
Reimplemented this as a handler. I'm not sure why the license and package.json changes are still showing up after I rebased.

File: src/handlers/yarn.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  console.log(`Fetch version data from registry.npmjs.org`);
+
+  const response = await fetch(`https://registry.npmjs.org/${module}`);
+  const data = await response.json();
+
+  return Object.keys(data.versions || {});
+}
+
+async function testCandidate(candidate, attempt, { module }) {
+  if (!originalVersion) {
```
Comment (iansu) [inline src/handlers/yarn.js:21]: I couldn't put this in `beforeAll` because I need the module name.

File: src/handlers/yarn.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import ora from 'ora';
+import fetch from 'isomorphic-fetch';
+import { exec, log } from '../util';
+
+let originalVersion;
+
+async function afterAll() {
+  await exec(`yarn add ${originalVersion} --no-lockfile`);
```
Comment (iansu) [inline src/handlers/yarn.js:8]: Unfortunately Yarn doesn't allow you to update a dependency without updating package.json so we have to put things back the way they were at the end.

File: src/handlers/yarn.js (hunk 3 above)
Comment (smeijer) [reply src/handlers/yarn.js:21]: We can of course also update the `beforeAll` call to include `args`? `beforeAll(args)`. What do you think?

File: src/index.js [hunk 5]
```diff
@@ -16,7 +17,9 @@ async function main(args) {
 
   if (interactive) {
     handler = handlers.manual;
-  } else if (args.module) {
+  } else if (args.module && fs.existsSync('yarn.lock')) {
+    handler = handlers.yarn;
+  } else {
     handler = handlers.npm;
   }
```
Comment (MichaelDeBoey) [inline src/index.js]:
```suggestion
  } else if (args.module) {
    if (fs.existsSync('yarn.lock')) {
      handler = handlers.yarn;
    } else {
      handler = handlers.npm;
    }
  }
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

File: src/handlers/yarn.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+async function testCandidate(candidate, attempt, { module }) {
+  if (!originalVersion) {
+    const output = JSON.parse(
+      (await exec(`yarn list --pattern ${module} --depth=0 --json`)).stdout
+        .replace("'", '')
+        .replace('\n', ''),
+    );
+    originalVersion = output.data.trees[0].name;
```
Comment (smeijer) [inline src/handlers/yarn.js:27]: Does this have a benefit over reading the version from `package.json`? We could even just backup the entire `package.json` file by copying it to a temp dir, and move it back once we're done.

Not saying this is wrong. I'm just curious.

File: src/index.js [hunk 7]
```diff
@@ -16,7 +17,9 @@ async function main(args) {
 
   if (interactive) {
     handler = handlers.manual;
-  } else if (args.module) {
+  } else if (args.module && fs.existsSync('yarn.lock')) {
+    handler = handlers.yarn;
+  } else {
```
Comment (smeijer) [inline src/index.js]:
```suggestion
  } else if (args.module) {
```

File: src/handlers/yarn.js (hunk 6 above)
Comment (iansu) [reply src/handlers/yarn.js:27]: No, I don't think there's any benefit to doing it this way. They both should produce the same result.

File: src/handlers/yarn.js (hunk 3 above)
Comment (iansu) [reply src/handlers/yarn.js:21]: Yeah, I also considered that. I'm happy to do that if you prefer.

File: src/index.js [hunk 8]
```diff
@@ -16,6 +17,8 @@ async function main(args) {
 
   if (interactive) {
     handler = handlers.manual;
+  } else if (args.module && fs.existsSync('yarn.lock')) {
+    handler = handlers.yarn;
   } else if (args.module) {
     handler = handlers.npm;
   }
```
Comment (MichaelDeBoey) [inline src/index.js:24]: Only check once for `args.module`

```suggestion
  } else if (Boolean(args.module))) {
    if (fs.existsSync('yarn.lock')) {
      handler = handlers.yarn;
    } else {
      handler = handlers.npm;
    }
  }
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 8 above)
Comment (smeijer) [reply src/index.js:24]: Actually, I prefer to check twice, and prevent the double nesting. Keep the structure flat.

**Review by smeijer — APPROVED** [review]

**Comment by smeijer** [comment]
@all-contributors please add @iansu for code

**Comment by smeijer** [comment]
Thanks for your work on this one. Much appreciated!

**Comment by iansu** [comment]
Happy to help. I've already used this tool to troubleshoot an issue with a dependency and it worked great!

**Pull request merged**